### PR TITLE
Fix RAT hangs by avoiding double free in TR_PatchNOPedGuardSiteOnClas…

### PR DESCRIPTION
**Pull request by fjeremic**
…sPreInitialize

When a class which has runtime assumptions registered against it gets
initialized (for example `Ljava/lang/String$StringCompressionFlag;`)
the `classGotInitialized` function will be invoked which will walk the
list of assumptions looking for the particular class signature.

If such an assumption is found `compensate` will be called so that the
runtime assumption patching logic is invoked. In the particular case of
the `TR_PatchNOPedGuardSiteOnClassPreInitialize` the compensation logic
will reclaim any persistent memory allocated when the assumption was
initialized. This particular class stores the full signature of class
the assumption is to trigger on in persistent memory, thus it will free
the allocated memory when `reclaim` is invoked inside of `compensate`.

The issue with this is that once the assumption is compensated it is
marked for removal. At a later point we attempt to reclaim the
assumption (in `reclaimMarkedAssumptionsFromRAT`) at which point
`reclaim` is invoked again on the runtime assumption. This second
reclaim will again deallocate the persistent memory thus causing a
double free to happen.

The observed issue is a hang in the RAT reclamation logic because in
between the two calls to `reclaim` another runtime assumption was
allocated in the JIT persistent memory which was freed for the class
signature. When the second `reclaim` is invoked we inadvertently free
the JIT persistent memory associated with another runtime assumption
and we end up in an undefined scenario which manifests itself as a hang.

To fix this issue we ensure no double free happens by validating the
allocated pointer before freeing and painting the pointer after freeing.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

### CODE REVIEW

**Comment by fjeremic** [comment]
Some details on how this bug was found for archeological purposes. This was a joint effort by @andrewcraik and @dsouzai. Many thanks for their help.

The hang was originally identified in a web server workload but was reproduced more recently in a contained unit test exercising a fix for #6671 on Z. In the unit test we had to enable `-XX:+CompactStrings` and this was the key option which made the hang much more reproducible.

The observed symptom was a hang in the RAT reclamation functions. There were several places observed to hang, but the most common had the following backtrace:

```
(gdb) bt
#0  0x000003ffa6051746 in OMR::RuntimeAssumption::dequeueFromListOfAssumptionsForJittedBody() () from /sandbox1/fjeremic/defects/6671/jvmxz6480/jre/lib/s390x/compressedrefs/libj9jit29.so
#1  0x000003ffa60522a2 in TR_RuntimeAssumptionTable::reclaimMarkedAssumptionsFromRAT(int)
...
```

It was determined that the reason for the hang was a loop in the doubly linked list of runtime assumptions for a particular method (without encountering the sentinel assumption). Thus when traversing the list of assumptions we would never reach the end. It was certainly non-obvious how the circularity was introduced, nor was it easy to find.

A `checkTable()` function was crafted to traverse the entire RAT. The hypothesis was that if we call this `checkTable()` function at various points and we hang inside it means the circularity was introduced. And indeed placing the call at every mutation of the RAT we were able to get a hang inside of `checkTable()`:

```
void TR_RuntimeAssumptionTable::checkTable()
{
    assumptionTableMutex->enter();

    if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseRuntimeAssumptions) && enableCheckTablePrinting)
    {
        TR_VerboseLog::writeLineLocked(TR_Vlog_RA, "Filip: Starting checkTable() printing");
    }

    for (int kind=0; kind < LastAssumptionKind; kind++) // for each table
    {
        TR_RatHT *hashTable = _tables + kind;
        for (size_t i = 0; i < hashTable->_spineArraySize; ++i) // for each bucket in the table
        {
            for (OMR::RuntimeAssumption *cursor = hashTable->_htSpineArray[i];  cursor; cursor = cursor->getNextEvenIfDead())
            {
                OMR::RuntimeAssumption *current = cursor->getNextAssumptionForSameJittedBodyEvenIfDead();
                if (current)
                {
                    while (current != cursor)
                    {
                        current = current->getNextAssumptionForSameJittedBodyEvenIfDead();

                        if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseRuntimeAssumptions) && enableCheckTablePrinting)
                        {
                            TR_VerboseLog::writeLineLocked(TR_Vlog_RA, "Performing Circularity Check for %p key = %p", current, current->getKey());
                        }
                    }
                }
            }
        }
    }

    if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseRuntimeAssumptions) && enableCheckTablePrinting)
    {
        TR_VerboseLog::writeLineLocked(TR_Vlog_RA, "Filip: Ending checkTable() printing");
    }
   assumptionTableMutex->exit();
}
```

From the verbose log we saw that the location we hang in was always inside of `TR_RuntimeAssumptionTable::addAssumption`, and in fact at the start of the function. Placing `checkTable()` calls at the removal points inside `reclaimMarkedAssumptionsFromRAT` was not triggering hangs. What was determined was that the hang in `checkTable()` did not happen after the last RAT cleanup and before the next addition, but it did happen when we attempted to add a new assumption. Since the RAT is not modified between the two calls to `checkTable()`, the latter of which was the one to hang, it must mean some sort of corruption was happening.

Taking a look at the verbose output and printing the entire RAT after the last cleanup and before the next addition we saw the following:

```
#RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43AC160: key=000003FFE43AC120 location=000003FFE5AE8012 destination=000003FFE5AE801E
#RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43AC0D0: key=000003FFE43AC090 location=000003FFE5AE8052 destination=000003FFE5AE801E
#RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43AC040: key=000003FFE43AC000 location=000003FFE5AE807E destination=000003FFE5AE8088
#RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43ABFB0: key=000003FFE43ABF70 location=000003FFE5AE80B4 destination=000003FFE5AE80BE
#RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43ABF20: key=000003FFE43ABEE0 location=000003FFE5AE810A destination=000003FFE5AE8114
#RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43ABE90: key=000003FFE43ABE50 location=000003FFE5AE8140 destination=000003FFE5AE814A
#RA:  Filip: Starting checkTable() printing
#RA:  Performing Circularity Check for 000003FFE43A9B50
#RA:  Performing Circularity Check for 000003FFE43A9AC0
#RA:  Performing Circularity Check for 000003FFE43A9A30
#RA:  Performing Circularity Check for 000003FFE43A99A0
...
#RA:  Performing Circularity Check for 000003FFE44370F0
#RA:  Performing Circularity Check for 000003FFE4437040
#RA:  Performing Circularity Check for 000003FFE43ABE50
#RA:  Performing Circularity Check for 000003FFE43ABE50
#RA:  Performing Circularity Check for 000003FFE43ABE50
#RA:  Performing Circularity Check for 000003FFE43ABE50
...
```

This was our hang, and the circular list. We also compared how the RAT looked before and after the hang, when no RAT mutation had taken place and we saw the following in the `vimdiff`:

```
+ +--72565 lines: #RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@|+ +--72565 lines: #RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@  #RA:  Performing Circularity Check for 000003FFE4437140                                     |  #RA:  Performing Circularity Check for 000003FFE4437140
  #RA:  Performing Circularity Check for 000003FFE44370F0                                     |  #RA:  Performing Circularity Check for 000003FFE44370F0
  #RA:  Performing Circularity Check for 000003FFE4437140                                     |  #RA:  Performing Circularity Check for 000003FFE4437140
  #RA:  Performing Circularity Check for 000003FFE44370F0                                     |  #RA:  Performing Circularity Check for 000003FFE44370F0
  #RA:  Performing Circularity Check for 000003FFE4437040                                     |  #RA:  Performing Circularity Check for 000003FFE4437040
  #RA:  Performing Circularity Check for 000003FFE43ABE50                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
  #RA:  Performing Circularity Check for 000003FFE43FDB20                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
  #RA:  Performing Circularity Check for 000003FFE43FDB20                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
  #RA:  Performing Circularity Check for 000003FFE43FDA70                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
  #RA:  Performing Circularity Check for 000003FFE43FD890                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
  #RA:  Performing Circularity Check for 000003FFE43ABDC0                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
  #RA:  Performing Circularity Check for 000003FFE43FD9E0                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
  #RA:  Performing Circularity Check for 000003FFE43FD990                                     |  #RA:  Performing Circularity Check for 000003FFE43ABE50
```

The most obviously incorrect thing was that the `key` of the last assumption we reclaimed was an assumption itself! i.e. this particular line:

```
#RA:  Filip Deleting ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43ABE90: key=000003FFE43ABE50 location=000003FFE5AE8140 destination=000003FFE5AE814A
#RA:  Filip: Starting checkTable() printing
```

has a `key` of `000003FFE43ABE50` which can be found in the RAT. This is the value we end up looping on.

There does not exist any runtime assumption which uses another assumption as a key, so this is obviously wrong. Now tracing back in the verbose log we can see where the original assumption associated with that key was added:

```
#RA:  Adding ClassPreInitialize assumption: TR::PatchNOPedGuardSite@000003FFE43ABE90: key=000003FFE43ABE50 location=000003FFE5AE8140 destination=000003FFE5AE814A
```

From this we can determine it is a `TR_PatchNOPedGuardSiteOnClassPreInitialize` runtime assumption. For sanity sake placing a breakpoint in the `TR_PatchNOPedGuardSiteOnClassPreInitialize::make` function we can see the key we store in the assumption is the class signature:

```
Thread 4 "JIT Compilation" hit Breakpoint 2, TR_PatchNOPedGuardSiteOnClassPreInitialize::make (fe=0x3ffe0007b90, persistentMemory=0x3fff80bc610, sig=0x3ffe4aa24a0 "Ljava/lang/String$StringCompressionFlag;", sigLen=40, loc=0x3ffe5a3b8a8 "\247\004\001$\030", dest=0x3ffe5a3baf0 "\030", sentinel=0x3ffe4933638) at /trjit/sandbox1/fjeremic/defects/6671/jvmxz6480/jre/lib/s390x/compressedrefs/compiler/..//compiler/runtime/RuntimeAssumptions.cpp:47
47         char * persistentSig = (char*)persistentMemory->allocatePersistentMemory(sizeof(char) * sigLen);

>>> p sig
$6 = 0x3ffe4aa24a0 "Ljava/lang/String$StringCompressionFlag;"
```

And the key is valid once the assumption is created. Randomly stopping in gdb before the hang point we can check the same pointer again and see that it is corrupted! The only way this can happen is if the assumption is no longer alive, i.e. it was reclaimed. And indeed in the `compensate` function of the assumption class we can see a call to `reclaim` which releases the persistent memory allocated to hold the signature (the `key` of the runtime assumption).

From here it is straightforward to determine that `reclaim` is called twice for this assumption, once when the `compensate` happens (when the first decompressed string is created) and the second time once the assumption is reclaimed.

Now the reason they `key` appeared in the RAT was precisely because the memory for the signature was freed from the persistent memory pool and the same memory was reused for another runtime assumption which was then added to the RAT. We then end up inadvertently freeing (calling `reclaim` during RAT reclamation) the `key` which now points to another runtime assumption and this is where the corruption happens which results in a circular list in the RAT for a method.

File: runtime/compiler/runtime/RuntimeAssumptions.cpp [hunk 1]
```diff
@@ -51,15 +51,6 @@ TR_PatchNOPedGuardSiteOnClassPreInitialize::make(
    return result;
    }
 
-
-void
-TR_PatchNOPedGuardSiteOnClassPreInitialize::compensate(TR_FrontEnd *fe, bool isSMP, void *data)
-   {
-   reclaim();
-   TR::PatchNOPedGuardSite::compensate(fe, isSMP, data);
-   }
```
Comment (fjeremic) [inline runtime/compiler/runtime/RuntimeAssumptions.cpp]: @andrewcraik @dsouzai it was not easily possible to leave this code in place. I assume the original intention of this code was to reclaim the persistent memory allocated for the class signature right after calling compensate as this is the earliest such a call can be made. However calling `reclaim` at this point makes things difficult for us because we are now zeroing out the `_key` after freeing the memory to ensure we don't double free.

Because the `_key` is zero following a call to `compensate` we can no longer calculate the hashcode of this assumption by calling the `hashcode` API [1], which is called from [2] after calling `compensate`. i.e. to mark the correct bucket in the RAT which needs to be reclaimed later we must calculate the hashcode [3], and to calculate the hashcode we need the class signature which we already freed.

The alternatives are to precalculate the hashcode and stash it off on the assumption object aggressively or to not free the class signature presistent memory and let the reclamation logic take care of it, which should happen shortly after compensating anyway.

In this PR I chose the second option as it is simpler and without any surprises. It is done at the cost of a very slight time window in which a small amount of persistent memory may be alive when it doesn't need to be. I think the complexity in this case is not worth the trade off for the space savings. However I'd like your opinion on this.

[1] https://github.com/eclipse/openj9/blob/347c34b062dba021da57e9539953f3a0714ea227/runtime/compiler/runtime/RuntimeAssumptions.cpp#L63-L91
[2] https://github.com/eclipse/openj9/blob/347c34b062dba021da57e9539953f3a0714ea227/runtime/compiler/runtime/RuntimeAssumptions.cpp#L366
[3] https://github.com/eclipse/openj9/blob/347c34b062dba021da57e9539953f3a0714ea227/runtime/compiler/runtime/ClassUnloadAssumption.cpp#L376-L383

**Comment by fjeremic** [comment]
Suggestion @andrewcraik , @dsouzai, and @mpirvu for review.

**Comment by 0xdaryl** [comment]
Kudos on the excellent writeup @fjeremic!

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 1 above)
Comment (andrewcraik) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp]: With the lazy reclamation the window for it to be freed is not guaranteed to be very small, but I think the more important thing is to fix the hangs and other issues definitively with a 'simple ' change and we can then try to optimize the memory management. Perhaps we could free up a bit in the hashcode with a left shift and the pointer field could point to the string or if low tagged could be interpreted as a hashcode left in place of the string. More complex for sure, but would probably achieve what we want in terms of memory management etc.

That being said, I am fine with this fix for now.

File: runtime/compiler/runtime/RuntimeAssumptions.cpp [hunk 2]
```diff
@@ -90,20 +81,34 @@ TR_PatchNOPedGuardSiteOnClassPreInitialize::hashCode(char *sig, uint32_t sigLen)
    return sum;
    }
 
+void
+TR_PatchNOPedGuardSiteOnClassPreInitialize::reclaim()
+   {
+   if (_key != 0)
```
Comment (andrewcraik) [inline runtime/compiler/runtime/RuntimeAssumptions.cpp]: Do we still need this since I think we are no longer nulling `_key` in reclaim the original code should work and might be less confusing - eg how could the key ever be null?

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 1 above)
Comment (dsouzai) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp]: The way I see it, the allocated string in `_key` is essentially part of the runtime assumption. It is already the case that assumptions in general are not immediately reclaimed after they fire. Therefore, we shouldn't treat the string allocation any different. By reclaiming all aspects of a given runtime assumption in one place at one time, it simplifies the code and prevents non-trivial bugs. 

tl;dr I'm fine with the fix as it stands.

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 2 above)
Comment (dsouzai) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp]: Might be worth doing a `TR_ASSERT_FATAL(_key != NULL, ....);`

File: runtime/compiler/env/CHTable.hpp [hunk 3]
```diff
@@ -127,10 +127,9 @@ class TR_PatchNOPedGuardSiteOnClassPreInitialize : public TR::PatchNOPedGuardSit
       TR_FrontEnd *fe, TR_PersistentMemory *, char *sig, uint32_t sigLen, uint8_t *loc, uint8_t *dest, OMR::RuntimeAssumption **sentinel);
    static uintptrj_t hashCode(char *sig, uint32_t sigLen);
 
-   virtual void reclaim() { jitPersistentFree((void*)_key); }
+   virtual void reclaim();
```
Comment (andrewcraik) [inline runtime/compiler/env/CHTable.hpp]: I am wondering if it would be wises for you to add a line or two of doxygen to note that the assumption persistent resources should only be freed in reclaim since the assumption can still be looked at/processed until the reclaim is run during deallocation. Might help make sure someone else doesn't make the same mistake in a compensate handler in the future?

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 1 above)
Comment (mpirvu) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp]: I am ok too with the solution chosen here.

File: runtime/compiler/runtime/RuntimeAssumptions.cpp (hunk 2 above)
Comment (fjeremic) [reply runtime/compiler/runtime/RuntimeAssumptions.cpp]: Fixed in [force push](https://github.com/eclipse/openj9/compare/f73530051f1fd15719269f628d2fe20c9065869f..5b57a0b92d691020cd06040de2ce7f1d86a74306).

File: runtime/compiler/env/CHTable.hpp (hunk 3 above)
Comment (fjeremic) [reply runtime/compiler/env/CHTable.hpp]: Fixed in [force push](https://github.com/eclipse/openj9/compare/f73530051f1fd15719269f628d2fe20c9065869f..5b57a0b92d691020cd06040de2ce7f1d86a74306). Note the base API lives in OMR. I've opened eclipse/omr#4256 to add the documentation there.

**Comment by fjeremic** [comment]
> LGTM, Might be worth adding to [/doc/compiler/BestPractices.md@`master`](https://github.com/eclipse/openj9/blob/master/doc/compiler/BestPractices.md) [](/eclipse/openj9/blob/HEAD@{2019-09-03T13:32:39Z}/doc/compiler/BestPractices.md)

I thought about this, however I'm not sure what exactly we would add. I suppose what we learned from this bug is to try to zero out JIT persistent memory after freeing to catch use after free bugs. Does this apply in general though as a best practice?

**Comment by fjeremic** [comment]
Only minor changes and documentation were added. Launching the builds preemptively here to speed up the process.

Jenkins test sanity all JDK8

**Comment by pshipton** [comment]
Jenkins test sanity all JDK8

**Comment by andrewcraik** [comment]
everyone agrees and the tests are good - merging. @fjeremic can you please open a PR to put this into the 0.16 branch as well?
